### PR TITLE
Proposal: Introduce new cop, Style/NoMetaprogramming, to discourage the use of meta programming

### PR DESCRIPTION
**Pull request by kellysutton**
This Pull Request defines a new cop, `Style/NoMetaprogramming`, that discourages the use of metaprogramming. 

As our engineering team has grown, we've seen some success with using Rubocop to nudge folks away from styles that we might no longer want to encourage within our application. Specifically, we've been removing metaprogramming from our business logic. We've found that this makes this easier to change the application and easier for newcomers to Ruby to understand most code.

I have set this to `default: false`, since I think it might be a bit controversial. Ruby's metaprogramming is what makes it so great, but we've found that it's best used in moderation.

I figure'd we'd upstream some of our work to give back to Rubocop. If this cop is too much, I don't mind closing the PR as well. Thanks for reviewing!

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/style/no_metaprogramming.rb [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        end
+
+        def on_def(node)
+          using_method_missing?(node) do
+            add_offense(node, message: METHOD_MISSING_MSG)
+          end
+        end
+
+        def on_send(node)
+          using_define_method?(node) do
```
Comment (marcandre) [inline lib/rubocop/cop/style/no_metaprogramming.rb:82]: Implementation wise, you could have a hash {method: 'error'} and this would be simpler and faster

**Comment by marcandre** [comment]
I ❤️  metaprogramming, thanks for proposing it to be disabled. I would consider also `Safe: false`.
You're missing a few callbacks (see `Lint/MissingSuper` for inspiration)

File: lib/rubocop/cop/style/no_metaprogramming.rb (hunk 1 above)
Comment (kellysutton) [reply lib/rubocop/cop/style/no_metaprogramming.rb]: Good idea. I've integrated that into subsequent commits.

**Comment by kellysutton** [comment]
@marcandre 

> I ❤️ metaprogramming, thanks for proposing it to be disabled.

Indeed, that's what makes Ruby so great. We use Rubocop as a tool for teaching newer engineers to Ruby what is okay vs. what might need closer consideration. Over the years, we've found metaprogramming to be one of those things that needs a double-check. But without metaprogramming, we wouldn't have Rails, RSpec, Rubocop, etc. :)

File: lib/rubocop/cop/style/no_metaprogramming.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      # This cop specifically will warn on usage of `.included`, `.inherited`,
+      # `#method_missing`, `.define_method`, `#instance_eval`, `.class_eval`,
+      # and `.define_singleton_method`.
+      #
+      class NoMetaprogramming < Base
+        def_node_matcher :using_method_missing?, <<~PATTERN
+          (def :method_missing ...)
+        PATTERN
+
+        METHOD_MISSING_MSG = 'Please do not use method_missing. Instead, explicitly define the'\
```
Comment (bbatsov) [inline lib/rubocop/cop/style/no_metaprogramming.rb:20]: I'd be nice if all your messages followed this format, as I see in some of them it's inverted. Let's avoid this inconsistency. Also - please put backticks around all method names referred in those messages. (e.g. `method_missing`).

File: lib/rubocop/cop/style/no_metaprogramming.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+module RuboCop
+  module Cop
+    module Style
+      # Discourages the use of metaprogramming. It's designed
+      # to be enabled in select folders within your application. It was
+      # originally developed with the idea of "Business logic should minimize
+      # the amount of metaprogramming."
+      #
+      # This cop specifically will warn on usage of `.included`, `.inherited`,
```
Comment (bbatsov) [inline lib/rubocop/cop/style/no_metaprogramming.rb:11]: Looking at the actual code it seems to disable a few more things which should be mentioned here. The cop also needs some good/bad code examples.

**Comment by bbatsov** [comment]
I'd name the cop something more neural - e.g. just Metaprogramming. Might be nice to have a certain degree of configuration as to what's allowed and not allowed, as I don't think all metaprogramming methods introduce a lot of complexity/mental overhead, but as the cop's disabled by default I'd be fine to accept it in its current state. You'll also have to rebase on top of master, as RuboCop 0.89 was released a while ago.

**Comment by bbatsov** [comment]
I'm closing this PR due to no recent activity. Feel free to re-open it if you ever come back to it.

**Pull request closed without merging**